build/source/driver/multi_driver.f90

```diff
@@ -369,69 +525,88 @@ real(dp),parameter        :: safety = 0.01_dp               ! safety factor to e
   ! write parameters for each HRU, and re-set indices
   do iHRU=1,nHRU
-   attr_data => attr_hru(iHRU)
-   type_data => type_hru(iHRU)
-   mpar_data => mpar_hru(iHRU)
-   indx_data => indx_hru(iHRU)
    ! write model parameters to the model output file
-   call writeAttrb(fileout,iHRU,err,message); call handle_err(err,message)
-   call writeParam(fileout,iHRU,err,message); call handle_err(err,message)
+   call writeAttrb(fileout,iHRU,attrStruct%hru(iHRU)%var,typeStruct%hru(iHRU)%var,err,message); call handle_err(err,message)
+   call writeParam(fileout,iHRU,mparStruct%hru(iHRU)%var,bparStruct%var,err,message); call handle_err(err,message)
    ! re-initalize the indices for midSnow, midSoil, midToto, and ifcToto
    jStep=1
-   indx_data%var(iLookINDEX%midSnowStartIndex)%dat(1) = 1
-   indx_data%var(iLookINDEX%midSoilStartIndex)%dat(1) = 1
-   indx_data%var(iLookINDEX%midTotoStartIndex)%dat(1) = 1
-   indx_data%var(iLookINDEX%ifcSnowStartIndex)%dat(1) = 1
-   indx_data%var(iLookINDEX%ifcSoilStartIndex)%dat(1) = 1
-   indx_data%var(iLookINDEX%ifcTotoStartIndex)%dat(1) = 1
+   indxStruct%hru(iHRU)%var(iLookINDEX%midSnowStartIndex)%dat(1) = 1
+   indxStruct%hru(iHRU)%var(iLookINDEX%midSoilStartIndex)%dat(1) = 1
+   indxStruct%hru(iHRU)%var(iLookINDEX%midTotoStartIndex)%dat(1) = 1
+   indxStruct%hru(iHRU)%var(iLookINDEX%ifcSnowStartIndex)%dat(1) = 1
+   indxStruct%hru(iHRU)%var(iLookINDEX%ifcSoilStartIndex)%dat(1) = 1
+   indxStruct%hru(iHRU)%var(iLookINDEX%ifcTotoStartIndex)%dat(1) = 1
   end do  ! (looping through HRUs)
  endif  ! if start of a new water year, and defining a new file
 
  ! initialize runoff variables
- bvar_data%var(iLookBVAR%basin__SurfaceRunoff)%dat(1)    = 0._dp  ! surface runoff (m s-1)
- bvar_data%var(iLookBVAR%basin__ColumnOutflow)%dat(1)    = 0._dp  ! outflow from all "outlet" HRUs (those with no downstream HRU)
+ bvarStruct%var(iLookBVAR%basin__SurfaceRunoff)%dat(1)    = 0._dp  ! surface runoff (m s-1)
+ bvarStruct%var(iLookBVAR%basin__ColumnOutflow)%dat(1)    = 0._dp  ! outflow from all "outlet" HRUs (those with no downstream HRU)
 
  ! initialize baseflow variables
- bvar_data%var(iLookBVAR%basin__AquiferRecharge)%dat(1)  = 0._dp ! recharge to the aquifer (m s-1)
- bvar_data%var(iLookBVAR%basin__AquiferBaseflow)%dat(1)  = 0._dp ! baseflow from the aquifer (m s-1)
- bvar_data%var(iLookBVAR%basin__AquiferTranspire)%dat(1) = 0._dp ! transpiration loss from the aquifer (m s-1)
+ bvarStruct%var(iLookBVAR%basin__AquiferRecharge)%dat(1)  = 0._dp ! recharge to the aquifer (m s-1)
+ bvarStruct%var(iLookBVAR%basin__AquiferBaseflow)%dat(1)  = 0._dp ! baseflow from the aquifer (m s-1)
+ bvarStruct%var(iLookBVAR%basin__AquiferTranspire)%dat(1) = 0._dp ! transpiration loss from the aquifer (m s-1)
 
  ! initialize total inflow for each layer in a soil column
  do iHRU=1,nHRU
-  mvar_hru(iHRU)%var(iLookMVAR%mLayerColumnInflow)%dat(:) = 0._dp
+  fluxStruct%hru(iHRU)%var(iLookFLUX%mLayerColumnInflow)%dat(:) = 0._dp
```

Comment:
OK - this is the stuff where I have to admit that fortran is nice. In C we'd have to loop over the soil layers ....